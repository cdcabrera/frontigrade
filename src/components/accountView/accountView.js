import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, Card, EmptyState, Grid, ListView, Modal, Row, Spinner } from 'patternfly-react';
import { Breadcrumb, BreadcrumbHeading } from '@patternfly/react-core';
import { withRouter } from 'react-router-dom';
import _isEqual from 'lodash/isEqual';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';
import apiTypes from '../../constants/apiConstants';
import accountViewTypes from './accountViewConstants';
import AccountGraphCard from '../accountGraphCard/accountGraphCard';
import ViewToolbar from '../viewToolbar/viewToolbar';
import AccountViewListItem from './accountViewListItem';

class AccountView extends React.Component {
  componentDidMount() {
    const { filter, getAccounts } = this.props;

    if (filter.query) {
      getAccounts(filter.query);
    }
  }

  componentDidUpdate(prevProps) {
    const { filter, getAccounts, updateAccounts } = this.props;

    if (
      (updateAccounts === true && updateAccounts !== prevProps.updateAccounts) ||
      !_isEqual(filter.query, prevProps.filter.query)
    ) {
      getAccounts(filter.query);
    }
  }

  onAddAccount = () => {
    store.dispatch({
      type: reduxTypes.account.ADD_ACCOUNT_SHOW
    });
  };

  onClearResetFilters = () => {
    const { view, viewGlobal } = this.props;

    store.dispatch({
      type: reduxTypes.filter.TOOLBAR_CLEAR_FILTERS,
      view,
      viewGlobal
    });
  };

  onDetailView = item => {
    const { history } = this.props;

    if (Number.parseInt(item[apiTypes.API_RESPONSE_ACCOUNTS_IMAGES], 10) > 0) {
      history.push(`/accounts/${item[apiTypes.API_RESPONSE_ACCOUNTS_ID]}`);
    } else {
      const displayName = item[apiTypes.API_RESPONSE_ACCOUNTS_NAME] || item[apiTypes.API_RESPONSE_ACCOUNTS_ID];

      store.dispatch({
        type: reduxTypes.toastNotifications.TOAST_ADD,
        alertType: 'info',
        message: (
          <span>
            No instances available for <strong>{displayName}</strong>
          </span>
        )
      });
    }
  };

  onDelete = account => {
    store.dispatch({
      type: reduxTypes.account.DELETE_ACCOUNT_SHOW,
      account
    });
  };

  onEditName = account => {
    store.dispatch({
      type: reduxTypes.account.EDIT_ACCOUNT_SHOW,
      account
    });
  };

  renderAccountsList() {
    const { accounts, filter, view, viewGlobal } = this.props;

    if (accounts.length) {
      const itemHasImagesClassName = item =>
        Number.parseInt(item[apiTypes.API_RESPONSE_ACCOUNTS_IMAGES], 10) > 0
          ? 'cloudmeter-accountview-list-view-item-active'
          : '';

      return (
        <React.Fragment>
          <AccountGraphCard filter={filter} view={view} viewGlobal={viewGlobal} />
          <Grid fluid>
            <Grid.Row className="row-cards-pf">
              <Grid.Col xs={12}>
                <Card className="cloudmeter-list-view-card">
                  <ListView className="cloudmeter-list-view">
                    {accounts.map(item => (
                      <AccountViewListItem
                        className={itemHasImagesClassName(item)}
                        filter={filter}
                        item={item}
                        key={item[apiTypes.API_RESPONSE_ACCOUNTS_ID]}
                        onDelete={this.onDelete}
                        onDetail={this.onDetailView}
                        onEdit={this.onEditName}
                      />
                    ))}
                  </ListView>
                </Card>
              </Grid.Col>
            </Grid.Row>
          </Grid>
        </React.Fragment>
      );
    }

    if (filter.activeFilters && filter.activeFilters.length) {
      return (
        <EmptyState className="list-view-blank-slate">
          <EmptyState.Title>No Results Match the Filter Criteria</EmptyState.Title>
          <EmptyState.Info>The active filters are hiding all items.</EmptyState.Info>
          <EmptyState.Action>
            <Button bsStyle="link" onClick={this.onClearResetFilters}>
              Clear Filters
            </Button>
          </EmptyState.Action>
        </EmptyState>
      );
    }

    return null;
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Modal bsSize="lg" backdrop={false} show animation={false}>
          <Modal.Body>
            <Spinner loading size="lg" className="blank-slate-pf-icon" />
            <div className="text-center">Loading...</div>
          </Modal.Body>
        </Modal>
      );
    }

    return null;
  }

  render() {
    const { accounts, error, errorMessage, filter, pending, view, viewGlobal } = this.props;

    if (error) {
      return (
        <EmptyState>
          <Alert type="error">
            <span>Error retrieving accounts: {errorMessage}</span>
          </Alert>
          {this.renderPendingMessage()}
        </EmptyState>
      );
    }

    if (pending && !accounts.length) {
      return <div className="cloudmeter-view-container">{this.renderPendingMessage()}</div>;
    }

    if (accounts.length || (filter.activeFilters && filter.activeFilters.length)) {
      return (
        <div className="cloudmeter-view-container fadein">
          <Grid fluid className="cloudmeter-app-header">
            <Breadcrumb className="cloudmeter-breadcrumb">
              <BreadcrumbHeading aria-current="page">Accounts</BreadcrumbHeading>
            </Breadcrumb>
          </Grid>
          <ViewToolbar
            dateFields={accountViewTypes.dateFields.timeValues}
            filterFields={accountViewTypes.filterFields}
            onAddAccount={this.onAddAccount}
            onExport={null}
            sortFields={accountViewTypes.sortFields}
            view={view}
            viewGlobal={viewGlobal}
            {...filter}
          />
          <div className="cloudmeter-list-container">{this.renderAccountsList()}</div>
          {this.renderPendingMessage()}
        </div>
      );
    }

    return (
      <Grid fluid>
        <Row>
          <EmptyState className="full-page-blank-slate fadein">
            <EmptyState.Icon />
            <EmptyState.Title>Welcome to Cloud Meter</EmptyState.Title>
            <EmptyState.Info>
              Cloud Meter collects usage analytics data for Red Hat cloud infrastructure products in your environments.
            </EmptyState.Info>
            <EmptyState.Info>Add an AWS account to get started.</EmptyState.Info>
            <EmptyState.Info>Adding an account includes creating a specialized IAM profile and role.</EmptyState.Info>
            <EmptyState.Info>
              For more information about the IAM steps in this process, see the{' '}
              <a
                href="https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html"
                target="_blank"
                rel="noopener noreferrer"
              >
                IAM User Guide
              </a>
              .
            </EmptyState.Info>
            <EmptyState.Action>
              <Button bsStyle="primary" bsSize="large" onClick={this.onAddAccount}>
                Add Account
              </Button>
            </EmptyState.Action>
          </EmptyState>
        </Row>
      </Grid>
    );
  }
}

AccountView.propTypes = {
  accounts: PropTypes.array,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  filter: PropTypes.shape({
    activeFilters: PropTypes.array,
    query: PropTypes.object
  }),
  getAccounts: PropTypes.func,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  pending: PropTypes.bool,
  updateAccounts: PropTypes.bool,
  view: PropTypes.string,
  viewGlobal: PropTypes.string
};

AccountView.defaultProps = {
  accounts: [],
  error: false,
  errorMessage: null,
  filter: {
    activeFilters: [],
    graphRhelValue: null,
    graphOpenshiftValue: null,
    query: {}
  },
  getAccounts: helpers.noop,
  pending: false,
  updateAccounts: false,
  view: 'account',
  viewGlobal: 'accountGlobal'
};

const mapDispatchToProps = dispatch => ({
  getAccounts: query => dispatch(reduxActions.account.getAccounts(query))
});

const mapStateToProps = state => ({
  ...state.account.view,
  filter: {
    ...state.filter.account,
    ...state.filter.accountGlobal,
    ...{
      query: {
        ...state.filter.account.query,
        ...state.filter.accountGlobal.query
      }
    }
  }
});

const ConnectedAccountView = withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(AccountView)
);

export { ConnectedAccountView as default, ConnectedAccountView, AccountView };
