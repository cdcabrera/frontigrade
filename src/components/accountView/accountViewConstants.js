import helpers from '../../common/helpers';

const dateFieldTypes = helpers.generatePriorYearMonthArray();

const filterFieldTypes = [
  {
    id: 'search_by_name',
    title: 'Name',
    placeholder: 'Filter by Name',
    filterType: 'text'
  }
];

const sortFieldTypes = [
  {
    id: 'name',
    title: 'Name',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'images',
    title: 'Images',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'instances',
    title: 'Instances',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'rhel',
    title: 'Rhel',
    isNumeric: false,
    sortAscending: true
  },
  {
    id: 'openshift',
    title: 'OpenShift',
    isNumeric: false,
    sortAscending: true
  }
];

const accountViewTypes = { dateFields: dateFieldTypes, filterFields: filterFieldTypes, sortFields: sortFieldTypes };

export { accountViewTypes as default, accountViewTypes, dateFieldTypes, filterFieldTypes, sortFieldTypes };
