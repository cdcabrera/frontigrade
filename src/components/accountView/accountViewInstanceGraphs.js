import React from 'react';
import PropTypes from 'prop-types';
import { Card, CardGrid, Label as PFLabel, Spinner, UtilizationCard } from 'patternfly-react';
import _sumBy from 'lodash/sumBy';
import { connect, reduxActions } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';

class AccountViewInstanceGraphs extends React.Component {
  static calculateInstanceTotals(dailyUsage = []) {
    const openshift = _sumBy(dailyUsage, val =>
      Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_OPENSHIFT_RUNTIME])
    );

    const rhel = _sumBy(dailyUsage, val => Number.parseFloat(val[apiTypes.API_RESPONSE_INSTANCES_USAGE_RHEL_RUNTIME]));

    return {
      openshiftHours: helpers.generateHoursFromSeconds(openshift),
      rhelHours: helpers.generateHoursFromSeconds(rhel)
    };
  }

  componentDidMount() {
    const { filter, getAccountInstances } = this.props;

    if (filter.query) {
      getAccountInstances(filter.query);
    }
  }

  componentDidUpdate(prevProps) {
    const { filter, getAccountInstances, updateInstances } = this.props;

    if (updateInstances === true && updateInstances !== prevProps.updateInstances) {
      getAccountInstances(filter.query);
    }
  }

  renderPendingCard() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Card matchHeight className="cloudmeter-utilization-graph cloudmeter-utilization-graph-loading fadein">
          <Card.Body>
            <Spinner loading size="sm" className="blank-slate-pf-icon" />
            <div className="text-center">
              <small>Loading...</small>
            </div>
          </Card.Body>
        </Card>
      );
    }

    return null;
  }

  render() {
    const { error, fulfilled, dailyUsage, instancesOpenshift, instancesRhel } = this.props;
    let displayTotals = {};

    if (error) {
      return null;
    }

    if (dailyUsage && dailyUsage.length) {
      displayTotals = { ...AccountViewInstanceGraphs.calculateInstanceTotals(dailyUsage) };
    }

    return (
      <CardGrid matchHeight>
        <CardGrid.Row>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Title>Red Hat Enterprise Linux</Card.Title>
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{instancesRhel}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      <PFLabel bsStyle="warning">
                        <abbr title="Red Hat Enterprise Linux">RHEL</abbr>
                      </PFLabel>{' '}
                      Instances
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Title>Red Hat OpenShift Container Platform</Card.Title>
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{instancesOpenshift}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      <PFLabel bsStyle="primary">
                        <abbr title="Red Hat OpenShift Container Platform">RHOCP</abbr>
                      </PFLabel>{' '}
                      Instances
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
          <CardGrid.Col sm={4}>
            {this.renderPendingCard()}
            {fulfilled && (
              <Card matchHeight className="cloudmeter-utilization-graph fadein">
                <Card.Body>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{displayTotals.rhelHours || 0}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>Red Hat Enterprise Linux Hours</UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                  <UtilizationCard.Details>
                    <UtilizationCard.DetailsCount>{displayTotals.openshiftHours || 0}</UtilizationCard.DetailsCount>
                    <UtilizationCard.DetailsDesc>
                      Red Hat OpenShift Container Platform Hours
                    </UtilizationCard.DetailsDesc>
                  </UtilizationCard.Details>
                </Card.Body>
              </Card>
            )}
          </CardGrid.Col>
        </CardGrid.Row>
      </CardGrid>
    );
  }
}

AccountViewInstanceGraphs.propTypes = {
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  filter: PropTypes.shape({
    query: PropTypes.object.isRequired
  }).isRequired,
  getAccountInstances: PropTypes.func,
  dailyUsage: PropTypes.array,
  instancesOpenshift: PropTypes.number,
  instancesRhel: PropTypes.number,
  pending: PropTypes.bool,
  updateInstances: PropTypes.bool
};

AccountViewInstanceGraphs.defaultProps = {
  error: false,
  fulfilled: false,
  getAccountInstances: helpers.noop,
  dailyUsage: [],
  instancesOpenshift: 0,
  instancesRhel: 0,
  pending: false,
  updateInstances: false
};

const mapDispatchToProps = dispatch => ({
  getAccountInstances: query => dispatch(reduxActions.account.getAccountInstances(query))
});

const mapStateToProps = state => ({
  ...state.account.instances
});

const ConnectedAccountViewInstanceGraphs = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountViewInstanceGraphs);

export { ConnectedAccountViewInstanceGraphs as default, ConnectedAccountViewInstanceGraphs, AccountViewInstanceGraphs };
