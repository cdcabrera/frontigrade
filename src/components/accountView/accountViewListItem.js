import React from 'react';
import PropTypes from 'prop-types';
import { DropdownKebab, Icon, ListView, Label as PFLabel, MenuItem } from 'patternfly-react';
import helpers from '../../common/helpers';

class AccountViewListItem extends React.Component {
  constructor(props) {
    super(props);
    this.kebab = React.createRef();
  }

  /**
   * FixMe: PF-React issue
   * Need a prop added to PF-React ListView for generic whole row event, instead of
   * having to use a "ref" and "contains" check
   */
  onVerifyDetail = event => {
    const { onDetail } = this.props;
    const currentKebabRef = this.kebab.current;

    if ((currentKebabRef && !currentKebabRef.contains(event.target)) || !currentKebabRef) {
      onDetail(event);
    }
  };

  renderActions() {
    const { item, onEdit, onArchive } = this.props;

    return (
      <div ref={this.kebab}>
        {(onEdit || onArchive) && (
          <DropdownKebab id={`account-item-menu-${item.id}`} pullRight>
            {onEdit && <MenuItem onClick={onEdit}>Edit Name</MenuItem>}
            {onArchive && <MenuItem onClick={onArchive}>Archive</MenuItem>}
          </DropdownKebab>
        )}
      </div>
    );
  }

  renderHeading() {
    const { item } = this.props;

    return <ListView.DescriptionHeading>{item.name}</ListView.DescriptionHeading>;
  }

  static renderLeftContent() {
    return null;
  }

  static renderDescription() {
    return null;
  }

  /**
   * ToDo: Eslint
   * Accessibility plugin tried to call out "Label" named component, converted to "PFLabel"
   * instead, look at Eslint settings
   */
  /**
   * FixMe: PF-React issue
   * listview "additionalInfo" attribute "requires" an array propType, restrictive limit.
   * Open it up to allow both, "node" OR "array"
   */
  static renderAdditionalInfo() {
    return [
      <ListView.InfoItem key="1">
        <Icon type="pf" name="cluster" />
        <strong>0</strong> Images
      </ListView.InfoItem>,
      <ListView.InfoItem key="2">
        <Icon type="pf" name="screen" />
        <strong>0</strong> Instances
      </ListView.InfoItem>,
      <ListView.InfoItem key="3">
        <strong>0</strong>
        <PFLabel bsStyle="primary">
          <abbr title="Red Hat Enterprise Linux">RHEL</abbr>
        </PFLabel>
      </ListView.InfoItem>,
      <ListView.InfoItem key="4">
        <strong>0</strong>
        <PFLabel bsStyle="primary">
          <abbr title="Red Hat Open Stack">RHOS</abbr>
        </PFLabel>
      </ListView.InfoItem>
    ];
  }

  render() {
    const { item } = this.props;

    return (
      <ListView.Item
        onClick={this.onVerifyDetail}
        key={item.id}
        leftContent={AccountViewListItem.renderLeftContent()}
        heading={this.renderHeading()}
        description={AccountViewListItem.renderDescription()}
        additionalInfo={AccountViewListItem.renderAdditionalInfo()}
        actions={this.renderActions()}
        stacked={false}
      />
    );
  }
}

AccountViewListItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string
  }).isRequired,
  onArchive: PropTypes.func,
  onDetail: PropTypes.func,
  onEdit: PropTypes.func
};

AccountViewListItem.defaultProps = {
  onArchive: null,
  onDetail: helpers.noop,
  onEdit: null
};

export { AccountViewListItem as default, AccountViewListItem };
