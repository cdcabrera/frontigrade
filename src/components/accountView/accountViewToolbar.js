import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'patternfly-react';
import { reduxTypes, store } from '../../redux';
import { fieldValidation } from '../formField/formField';
import accountViewTypes from './accountViewConstants';
import Toolbar from '../toolbar/toolbar';

class AccountViewToolbar extends React.Component {
  componentDidMount() {
    const { dateFields, dateValue, filterField, filterFields, sortFields, sortValue } = this.props;

    if (!dateValue) {
      this.onUpdateDateField(dateFields[0]);
    }

    if (!filterField) {
      this.onSelectFilter(filterFields[0]);
    }

    if (!sortValue) {
      this.onUpdateSortField(sortFields[0]);
    }
  }

  onAddAccount = () => {
    store.dispatch({
      type: reduxTypes.account.ADD_ACCOUNT_SHOW
    });
  };

  onExport = () => {
    store.dispatch({
      type: reduxTypes.toastNotifications.TOAST_ADD,
      alertType: 'warning',
      message: 'Export not yet enabled'
    });
  };

  onUpdateDateField = dateValue => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_SET_DATE_TYPE, { dateValue });
  };

  onClearFilters = () => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_CLEAR_FILTERS);
  };

  onRemoveFilter = filter => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_REMOVE_FILTER, { filter });
  };

  onSelectFilter = filterField => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_SET_FILTER_TYPE, { filterField });
  };

  onToggleSortDirection = () => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_TOGGLE_SORT_ASCENDING);
  };

  onUpdateSortField = sortValue => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_SET_SORT_TYPE, { sortValue });
  };

  onUpdateFilterValue = event => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_SET_FILTER_VALUE, { filterValue: event.target.value });
  };

  onKeyPress = (key, filterValue) => {
    if (key === 'Enter') {
      this.onAddFilterValue(filterValue);
    }
  };

  onSelectFilterValue = filterValue => {
    this.onDispatch(reduxTypes.filter.TOOLBAR_SET_FILTER_VALUE, { filterValue });
    this.onAddFilterValue(filterValue);
  };

  onAddFilterValue = filterValue => {
    const { filterField } = this.props;

    if (!fieldValidation.isEmpty(filterValue)) {
      const filterText = `${filterField.title || filterField}: ${filterValue.title || filterValue}`;

      this.onDispatch(reduxTypes.filter.TOOLBAR_ADD_FILTER, {
        filter: { field: filterField, value: filterValue, label: filterText }
      });
    }
  };

  onDispatch = (type, data = {}) => {
    const { view } = this.props;

    store.dispatch({
      type,
      view,
      ...data
    });
  };

  render() {
    const {
      activeFilters,
      dateValue,
      dateFields,
      filterField,
      filterFields,
      filterValue,
      resultsType,
      resultsTypePlural,
      sortAscending,
      sortFields,
      sortValue,
      selectedCount,
      totalCount
    } = this.props;

    return (
      <Toolbar>
        <Toolbar.Sort onUpdateField={this.onUpdateDateField} sortFields={dateFields} sortValue={dateValue} />
        <Toolbar.Filter
          filterField={filterField}
          filterFields={filterFields}
          filterValue={filterValue}
          onFilterSelected={this.onSelectFilterValue}
          onKeyPress={this.onKeyPress}
          onUpdateValue={this.onUpdateFilterValue}
          onSelectFilter={this.onSelectFilter}
        />
        <Toolbar.Sort
          onUpdateField={this.onUpdateSortField}
          onToggleDirection={this.onToggleSortDirection}
          sortAscending={sortAscending}
          sortFields={sortFields}
          sortValue={sortValue}
        />
        <Toolbar.Right>
          <Toolbar.Group>
            <Button bsStyle="primary" disabled={false} onClick={this.onAddAccount}>
              Add Account
            </Button>
            <Button disabled onClick={this.onExport}>
              Export
            </Button>
          </Toolbar.Group>
        </Toolbar.Right>
        <Toolbar.Results
          activeFilters={activeFilters}
          alignRight
          onClearFilters={this.onClearFilters}
          onRemoveFilter={this.onRemoveFilter}
          resultsType={resultsType}
          resultsTypePlural={resultsTypePlural}
          selectedCount={selectedCount}
          totalCount={totalCount}
        />
      </Toolbar>
    );
  }
}

AccountViewToolbar.propTypes = {
  activeFilters: PropTypes.array,
  dateValue: PropTypes.shape({
    end: PropTypes.string,
    start: PropTypes.string,
    title: PropTypes.string,
    value: PropTypes.string
  }),
  dateFields: PropTypes.array,
  filterField: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    placeholder: PropTypes.string,
    filterType: PropTypes.string,
    filterValues: PropTypes.array
  }),
  filterFields: PropTypes.array,
  filterValue: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  resultsType: PropTypes.string,
  resultsTypePlural: PropTypes.string,
  selectedCount: PropTypes.number,
  sortAscending: PropTypes.bool,
  sortFields: PropTypes.array,
  sortValue: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    isNumeric: PropTypes.bool,
    sortAscending: PropTypes.bool
  }),
  totalCount: PropTypes.number,
  view: PropTypes.string
};

AccountViewToolbar.defaultProps = {
  activeFilters: [],
  dateFields: accountViewTypes.dateFields,
  dateValue: null,
  filterField: null,
  filterFields: accountViewTypes.filterFields,
  filterValue: '',
  resultsType: 'Result',
  resultsTypePlural: 'Results',
  selectedCount: 0,
  sortAscending: true,
  sortFields: accountViewTypes.sortFields,
  sortValue: null,
  totalCount: 0,
  view: 'account'
};

export { AccountViewToolbar as default, AccountViewToolbar };
