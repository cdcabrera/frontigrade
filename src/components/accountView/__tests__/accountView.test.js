import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedAccountView, AccountView } from '../accountView';

describe('AccountView Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a connected component', () => {
    const store = generateEmptyStore({ account: {} });
    const component = shallow(<ConnectedAccountView />, { context: { store } });

    expect(component.dive()).toMatchSnapshot('connected');
  });

  it('should render a non-connected component', () => {
    const props = {};

    const component = mount(<AccountView {...props} />);
    expect(component.render()).toMatchSnapshot();

    expect(component.find('button').length).toEqual(1);
  });
});
