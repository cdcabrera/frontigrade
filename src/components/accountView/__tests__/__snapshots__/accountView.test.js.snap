// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountView Component should render a connected component with default props: connected 1`] = `
<Route>
  <Component />
</Route>
`;

exports[`AccountView Component should render a non-connected component in pending state: pending 1`] = `
<div
  class="cloudmeter-view-container"
/>
`;

exports[`AccountView Component should render a non-connected component in pending state: pending with list 1`] = `
<div
  className="cloudmeter-view-container fadein"
>
  <Grid
    bsClass="container"
    componentClass="div"
    fluid={true}
  >
    <Breadcrumb
      className="cloudmeter-breadcrumb"
      title={true}
    >
      <BreadcrumbItem
        active={true}
        aria-current="page"
      >
        <strong>
          Accounts
        </strong>
      </BreadcrumbItem>
    </Breadcrumb>
  </Grid>
  <ViewToolbar
    activeFilters={Array []}
    dateFields={null}
    dateValue={null}
    filterField={null}
    filterFields={
      Array [
        Object {
          "filterType": "text",
          "id": "name_pattern",
          "placeholder": "Filter by Name",
          "title": "Name",
        },
      ]
    }
    filterValue=""
    onAddAccount={[Function]}
    onExport={null}
    query={Object {}}
    resultsType="Result"
    resultsTypePlural="Results"
    selectedCount={0}
    sortAscending={true}
    sortFields={
      Array [
        Object {
          "id": "name",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Name",
        },
        Object {
          "id": "images",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Images",
        },
        Object {
          "id": "instances",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Instances",
        },
        Object {
          "id": "rhel",
          "isNumeric": false,
          "sortAscending": true,
          "title": "Rhel",
        },
        Object {
          "id": "openshift",
          "isNumeric": false,
          "sortAscending": true,
          "title": "OpenShift",
        },
      ]
    }
    sortValue={null}
    totalCount={0}
    view="account"
    viewGlobal="accountGlobal"
  />
  <div
    className="cloudmeter-list-container"
  >
    <Connect(AccountGraphCard)
      filter={
        Object {
          "dateFields": null,
          "query": Object {},
        }
      }
      view="account"
    />
    <Grid
      bsClass="container"
      componentClass="div"
      fluid={true}
    >
      <Row
        bsClass="row"
        className="row-cards-pf"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          xs={12}
        >
          <Card
            accented={false}
            aggregated={false}
            aggregatedMini={false}
            cardRef={null}
            className="cloudmeter-list-view-card"
            matchHeight={false}
          >
            <ListView
              className="cloudmeter-list-view"
            >
              <AccountViewListItem
                item={
                  Object {
                    "arn": "arn:aws:iam::TEST",
                    "creation_date": "2018-07-06T15:09:21.442412Z",
                    "id": "1",
                    "images": 1,
                    "instances": 2,
                    "name": "Lorem test",
                    "openshift_instances": null,
                    "rhel_instances": 1,
                    "type": "aws",
                    "user_id": 1,
                  }
                }
                key="1"
                onDelete={[Function]}
                onDetail={[Function]}
                onEdit={[Function]}
                t={[Function]}
              />
            </ListView>
          </Card>
        </Col>
      </Row>
    </Grid>
  </div>
  <Modal
    animation={false}
    autoFocus={true}
    backdrop={false}
    bsClass="modal"
    bsSize="lg"
    dialogComponentClass={[Function]}
    enforceFocus={true}
    keyboard={true}
    manager={
      ModalManager {
        "add": [Function],
        "containers": Array [],
        "data": Array [],
        "handleContainerOverflow": true,
        "hideSiblingNodes": true,
        "isTopModal": [Function],
        "modals": Array [],
        "remove": [Function],
      }
    }
    onHide={[Function]}
    renderBackdrop={[Function]}
    restoreFocus={true}
    show={true}
  >
    <ModalBody
      bsClass="modal-body"
      componentClass="div"
    >
      <Spinner
        className="blank-slate-pf-icon"
        inline={false}
        inverse={false}
        loading={true}
        size="lg"
      />
      <div
        className="text-center"
      >
        Loading...
      </div>
    </ModalBody>
  </Modal>
</div>
`;

exports[`AccountView Component should render a non-connected component with empty state: empty state 1`] = `
<div
  class="container-fluid"
>
  <div
    class="row"
  >
    <div
      class="blank-slate-pf full-page-blank-slate fadein"
    >
      <div
        class="blank-slate-pf-icon"
      >
        <span
          aria-hidden="true"
          class="pficon pficon-add-circle-o"
        />
      </div>
      <h4
        class="h1 blank-slate-pf-title"
      >
        Welcome to Cloud Meter
      </h4>
      <p
        class="blank-slate-pf-info"
      >
        Cloud Meter collects usage analytics data for Red Hat cloud infrastructure products in your environments.
      </p>
      <p
        class="blank-slate-pf-info"
      >
        Add an AWS account to get started.
      </p>
      <p
        class="blank-slate-pf-info"
      >
        Adding an account includes creating a specialized IAM profile and role.
      </p>
      <p
        class="blank-slate-pf-info"
      >
        For more information about the IAM steps in this process, see the 
        <a
          href="https://docs.aws.amazon.com/IAM/latest/UserGuide/introduction.html"
          rel="noopener noreferrer"
          target="_blank"
        >
          IAM User Guide
        </a>
        .
      </p>
      <div
        class="blank-slate-pf-main-action"
      >
        <button
          class="btn btn-lg btn-primary"
          type="button"
        >
          Add Account
        </button>
      </div>
    </div>
  </div>
</div>
`;
